pkg/virt-launcher/virtwrap/network/dhcp/filter.go

```diff
@@ -0,0 +1,68 @@
+// This file is inspired by github.com/krolaw/dhcp4/conn/filter.go.
+// Couldn't use the original one since SO_REUSEADDR option has to be set.
+package dhcp
+
+import (
+	"context"
+	"net"
+	"syscall"
+
+	"golang.org/x/net/ipv4"
+)
+
+// Creates listener on all interfaces and then filters packets not received by interfaceName
+func NewUDP4FilterListener(interfaceName, laddr string) (c *serveIfConn, e error) {
+	iface, err := net.InterfaceByName(interfaceName)
+	if err != nil {
+		return nil, err
+	}
+	lc := net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			var opErr error
+			err := c.Control(func(fd uintptr) {
+				opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
+			})
+			if err != nil {
+				return err
+			}
+			return opErr
+		},
+	}
+	l, err := lc.ListenPacket(context.Background(), "udp4", laddr)
+
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if e != nil {
+			l.Close()
+		}
+	}()
+	p := ipv4.NewPacketConn(l)
+	if err := p.SetControlMessage(ipv4.FlagInterface, true); err != nil {
+		return nil, err
+	}
+	return &serveIfConn{ifIndex: iface.Index, conn: p}, nil
+}
+
+type serveIfConn struct {
+	ifIndex int
+	conn    *ipv4.PacketConn
+	cm      *ipv4.ControlMessage
+}
+
+func (s *serveIfConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
+	for { // Filter all other interfaces
+		n, s.cm, addr, err = s.conn.ReadFrom(b)
```

Comment:
I find it "dangerous" to assign `s.cm` before assuring the read succeeded.
It changes the object state as a side effect and the caller may still perform a `WriteTo` even though the value is not populated correctly (e.g. it matched `s.cm == nil`).

In the condition below, if the `s.cm` has no value, it returns without an error and without a proper filtering.

Reply:
Changed to use `dhcpConn.NewServeIf`, so now I cannot change the method.